test/test_autograd.py

```diff
@@ -4129,6 +4130,877 @@ def fn(*inputs):
+                self.assertEqual(el_base.size(), el_res.size())
+        else:
+            # Wrong base
+            raise RuntimeError("The base given to `_assert_same_struct` doesn't have"
+                               " the right structure.")
+
+    def _assert_interleaved_struct(self, res, base1, base2):
+        # base1 and base2 can be Tensors or tuples of Tensors.
+        # If they are tuples, res should be a tuple as well.
+        # The indexing works as follow for base1, base2 being
+        # - tuple, tuple: res[i][j][k][l] = (base1[i][k], base2[j][l])
+        # - tuple, Tensor: res[i][k][l] = (base1[i][k], base2[l])
+        # - Tensor, tuple: res[i][j][l] = (base1[i], base2[j][l])
+        # - Tensor, Tensor: res[k][l] = (base1[k], base2[l])
+        if torch.is_tensor(base1) and torch.is_tensor(base2):
+            self.assertTrue(torch.is_tensor(res))
+            self.assertEqual(res.size(), base1.size() + base2.size())
+        elif isinstance(base1, tuple) and torch.is_tensor(base2):
+            self.assertTrue(isinstance(res, tuple))
+            self.assertEqual(len(res), len(base1))
+            for el_res, el_base1 in zip(res, base1):
+                self.assertTrue(torch.is_tensor(el_res))
+                self.assertTrue(torch.is_tensor(el_base1))
+                self.assertEqual(el_res.size(), el_base1.size() + base2.size())
+        elif torch.is_tensor(base1) and isinstance(base2, tuple):
+            self.assertTrue(isinstance(res, tuple))
+            self.assertEqual(len(res), len(base2))
+            for el_res, el_base2 in zip(res, base2):
+                self.assertTrue(torch.is_tensor(el_res))
+                self.assertTrue(torch.is_tensor(el_base2))
+                self.assertEqual(el_res.size(), base1.size() + el_base2.size())
+        elif isinstance(base1, tuple) and isinstance(base2, tuple):
+            self.assertTrue(isinstance(res, tuple))
+            self.assertEqual(len(res), len(base1))
+            for el_res, el_base1 in zip(res, base1):
+                self.assertTrue(isinstance(el_res, tuple))
+                self.assertEqual(len(res), len(base2))
+                for el_el_res, el_base2 in zip(el_res, base2):
+                    self.assertTrue(torch.is_tensor(el_el_res))
+                    self.assertTrue(torch.is_tensor(el_base2))
+                    self.assertEqual(el_el_res.size(), el_base1.size() + el_base2.size())
+        else:
+            # Wrong bases
+            raise RuntimeError("The bases given to `_assert_interleaved_struct` don't have"
+                               " the right structure.")
+
+    def test_vjp_err_check(self):
+        def foo(a):
+            return 3 * a.narrow(0, 0, 3)
+
+        def bar(a):
+            return 3 * a.narrow(0, 0, 3), "bar"
+
+        inp = torch.rand(4)
+        with self.assertRaisesRegex(TypeError, "The inputs given to vjp must be either a Tensor"):
+            res = autogradF.vjp(foo, (inp, 2))
```

Comment:
I can see someone "reversing" the v-check and the inputs check in the future. Should we pass in a correct `v` here to make it clear that we are testing incorrect inputs?

Reply:
Good point. Will do !

Reply:
Err actually, you can't "reverse" the checks because we need to check that the shape of `v` matches the output. The other point still stands, though, it might be nice to pass in a valid v here to make it clear what we're testing

Reply:
True for vjp. But not for jvp and hvp. I should at least update these two.